TestBadger still flaking

https://github.com/ssbc/go-ssb/actions/runs/3726907359/jobs/6320652343

> 2022/12/18 22:38:59 [margaret/indexes/badger] overwrote batch limit 0
> --- FAIL: TestBadger (0.15s)
>     --- FAIL: TestBadger/scene1 (0.11s)
>         builder_test.go:234: 
>             	Error Trace:	/home/runner/work/go-ssb/go-ssb/graph/builder_test.go:234
>             	Error:      	Not equal: 
>             	            	expected: 3
>             	            	actual  : 2
>             	Test:       	TestBadger/scene1
>     builder_test.go:131: closed scenary
> FAIL
> FAIL	github.com/ssbc/go-ssb/graph	0.159s
> FAIL
> make: *** [Makefile:22: test] Error 1
> Error: Process completed with exit code 2.

--- a/graph/builder.go
+++ b/graph/builder.go
@@ -59,6 +59,8 @@ type BadgerBuilder struct {
 	idxSinkMetaFeeds     librarian.SinkIndex
 	idxSinkAnnouncements librarian.SinkIndex
 
+	idxInSync   sync.WaitGroup
+
 	log log.Logger
 
 	cacheLock   sync.Mutex
@@ -86,6 +88,7 @@ func NewBuilder(log log.Logger, db *badger.DB, hmacSecret *[32]byte) *BadgerBuil
 }
 
 func (b *BadgerBuilder) DeleteAuthor(who refs.FeedRef) error {
+	b.WaitUntilIndexesAreSynced()
 	b.cacheLock.Lock()
 	defer b.cacheLock.Unlock()
 	b.cachedGraph = nil
@@ -116,6 +119,7 @@ func (b *BadgerBuilder) Authorizer(from refs.FeedRef, maxHops int) ssb.Authorize
 }
 
 func (b *BadgerBuilder) Build() (*Graph, error) {
+	b.WaitUntilIndexesAreSynced()
 	dg := NewGraph()
 
 	b.cacheLock.Lock()
@@ -244,6 +248,7 @@ func (l Lookup) Dist(to refs.FeedRef) ([]graph.Node, float64) {
 }
 
 func (b *BadgerBuilder) Follows(forRef refs.FeedRef) (*ssb.StrFeedSet, error) {
+	b.WaitUntilIndexesAreSynced()
 	fs := ssb.NewFeedSet(50)
 	err := b.kv.View(func(txn *badger.Txn) error {
 		iter := txn.NewIterator(badger.DefaultIteratorOptions)
@@ -283,6 +288,7 @@ func (b *BadgerBuilder) Follows(forRef refs.FeedRef) (*ssb.StrFeedSet, error) {
 
 // Metafeed returns the metafeed for a subfeed, or an error if it has none.
 func (b *BadgerBuilder) Metafeed(subfeed refs.FeedRef) (refs.FeedRef, error) {
+	b.WaitUntilIndexesAreSynced()
 	var found refs.FeedRef
 	err := b.kv.View(func(txn *badger.Txn) error {
 		iter := txn.NewIterator(badger.DefaultIteratorOptions)
@@ -316,6 +322,7 @@ func (b *BadgerBuilder) Metafeed(subfeed refs.FeedRef) (refs.FeedRef, error) {
 
 // Subfeeds returns the set of subfeeds for a particular metafeed.
 func (b *BadgerBuilder) Subfeeds(metaFeed refs.FeedRef) (*ssb.StrFeedSet, error) {
+	b.WaitUntilIndexesAreSynced()
 	fs := ssb.NewFeedSet(50)
 	err := b.kv.View(func(txn *badger.Txn) error {
 		iter := txn.NewIterator(badger.DefaultIteratorOptions)
@@ -361,6 +368,7 @@ func (b *BadgerBuilder) Subfeeds(metaFeed refs.FeedRef) (*ssb.StrFeedSet, error)
 //
 // See hops_test.go for concrete examples.
 func (b *BadgerBuilder) Hops(from refs.FeedRef, max int) *ssb.StrFeedSet {
+	b.WaitUntilIndexesAreSynced()
 	max++
 	walked := ssb.NewFeedSet(0)
 	visited := make(map[string]struct{}) // tracks the nodes we already recursed from (so we don't do them multiple times on common friends)

--- a/graph/builder_indexing.go
+++ b/graph/builder_indexing.go
@@ -7,6 +7,7 @@ package graph
 import (
 	"context"
 	"fmt"
+	"time"
 
 	"github.com/ssbc/go-ssb-refs/tfk"
 	"github.com/ssbc/margaret"
@@ -31,8 +32,27 @@ const (
 	idxRelValueMetafeed
 )
 
+func (b *BadgerBuilder) indexSyncStart() {
+	b.idxInSync.Add(1)
+}
+
+func (b *BadgerBuilder) indexSyncDone() {
+	// this delay is here so that the WaitGroup is held while Luigi continues to process more data
+	// TODO: eliminate this delay once we have a way to query Luigi directly to see if it's done with its source queue
+	time.AfterFunc(100 * time.Millisecond, func() {
+		b.idxInSync.Done()
+	})
+}
+
+// WaitUntilIndexesAreSynced blocks until all the index processing is in sync with the rootlog
+func (b *BadgerBuilder) WaitUntilIndexesAreSynced() {
+	b.idxInSync.Wait()
+}
+
 func (b *BadgerBuilder) updateAnnouncement(ctx context.Context, seq int64, val interface{}, idx librarian.SetterIndex) error {
 	b.cacheLock.Lock()
+	b.indexSyncStart()
+	defer b.indexSyncDone()
 	defer b.cacheLock.Unlock()
 
 	if nulled, ok := val.(error); ok {
@@ -72,6 +92,8 @@ func (b *BadgerBuilder) updateAnnouncement(ctx context.Context, seq int64, val i
 }
 
 func (b *BadgerBuilder) OpenAnnouncementIndex() (librarian.SeqSetterIndex, librarian.SinkIndex) {
+	b.indexSyncStart()
+	defer b.indexSyncDone()
 	if b.idxSinkAnnouncements == nil {
 		b.idxSinkAnnouncements = librarian.NewSinkIndex(b.updateAnnouncement, b.idx)
 	}
@@ -80,6 +102,8 @@ func (b *BadgerBuilder) OpenAnnouncementIndex() (librarian.SeqSetterIndex, libra
 
 func (b *BadgerBuilder) updateContacts(ctx context.Context, seq int64, val interface{}, idx librarian.SetterIndex) error {
 	b.cacheLock.Lock()
+	b.indexSyncStart()
+	defer b.indexSyncDone()
 	defer b.cacheLock.Unlock()
 
 	if nulled, ok := val.(error); ok {
@@ -136,6 +160,8 @@ func (b *BadgerBuilder) OpenContactsIndex() (librarian.SeqSetterIndex, librarian
 
 func (b *BadgerBuilder) updateMetafeeds(ctx context.Context, seq int64, val interface{}, idx librarian.SetterIndex) error {
 	b.cacheLock.Lock()
+	b.indexSyncStart()
+	defer b.indexSyncDone()
 	defer b.cacheLock.Unlock()
 
 	if nulled, ok := val.(error); ok {
@@ -259,6 +285,8 @@ func (b *BadgerBuilder) updateMetafeeds(ctx context.Context, seq int64, val inte
 }
 
 func (b *BadgerBuilder) OpenMetafeedsIndex() (librarian.SeqSetterIndex, librarian.SinkIndex) {
+	b.indexSyncStart()
+	defer b.indexSyncDone()
 	if b.idxSinkMetaFeeds == nil {
 		b.idxSinkMetaFeeds = librarian.NewSinkIndex(b.updateMetafeeds, b.idx)
 	}

--- a/graph/builder_test.go
+++ b/graph/builder_test.go
@@ -14,7 +14,6 @@ import (
 	"path/filepath"
 	"sync"
 	"testing"
-	"time"
 
 	"github.com/dgraph-io/badger/v3"
 	"github.com/ssbc/go-luigi"
@@ -134,11 +133,6 @@ func makeBadger(t *testing.T) testStore {
 }
 
 func TestBadger(t *testing.T) {
-	if testutils.SkipOnCI(t) {
-		// https://github.com/ssbc/go-ssb/pull/167
-		return
-	}
-
 	tc := makeBadger(t)
 	t.Run("scene1", tc.theScenario)
 }
@@ -232,11 +226,10 @@ func (tc testStore) theScenario(t *testing.T) {
 	myself.follow(alice.key.ID())
 	myself.block(bob.key.ID())
 
-	time.Sleep(time.Second / 10)
-
+	// this should result in 3 nodes - alice, bob, and myself
 	g, err = tc.gbuilder.Build()
 	r.NoError(err)
-	if !a.Equal(3, g.NodeCount()) {
+	if !a.Equal(3, g.NodeCount(), "wrong count for number of nodes in the graph") {
 		return
 	}
 
@@ -265,8 +258,6 @@ func (tc testStore) theScenario(t *testing.T) {
 		t.Fatal("please 'export LIBRARIAN_WRITEALL=0' for this test to pass")
 		// TODO: expose index flushing
 	}
-
-	time.Sleep(time.Second / 10)
 
 	g, err = tc.gbuilder.Build()
 	r.NoError(err)

